packages/core/contracts/merkle-distributor/implementation/MerkleDistributor.sol

```diff
@@ -0,0 +1,236 @@
+    // contract. Importantly, we assume that the owner of this contract
+    // correctly chooses an amount `totalRewardsDistributed` that is sufficient
+    // to cover all claims within the `merkleRoot`. Otherwise, a race condition
+    // can be created. This situation can occur because we do not segregate reward balances by window,
+    // for code simplicity purposes.
+    //
+    // Example race situation:
+    //     - Window 1 Tree: Owner sets `totalRewardsDistributed=100` and insert proofs that give
+    //                      claimant A 50 tokens and claimant B 51 tokens. The owner has made an error
+    //                      by not setting the `totalRewardsDistributed` correctly to 101).
+    //     - Window 2 Tree: Owner sets `totalRewardsDistributed=1` and insert proofs that give
+    //                      claimant A 1 token. The owner correctly set `totalRewardsDistributed` this time.
+    //     - At this point contract owns 100 + 1 = 101 tokens. Now, imagine the following sequence:
+    //       (1) Claimant A claims 50 tokens for Window 1, contract now has 101 - 50 = 51 tokens.
+    //       (2) Claimant B claims 51 tokens for Window 1, contract now has 51 - 51 = 0 tokens.
+    //       (3) Claimant A tries to claim 1 token for Window 2 but fails because contract has 0 tokens.
+    //     - In summary, the contract owner created a race for step(2) and step(3) in which the first
+    //       claim would succeed and the second claim would fail, even though both claimants would expect
+    //       their claims to suceed.
+    function setWindowMerkleRoot(
+        uint256 totalRewardsDistributed,
+        uint256 windowStart,
+        address rewardToken,
+        bytes32 merkleRoot
+    ) external nonReentrant() onlyOwner {
+        uint256 indexToSeed = lastSeededIndex;
+        lastSeededIndex = indexToSeed.add(1);
+
+        _seedWindow(indexToSeed, totalRewardsDistributed, windowStart, rewardToken, merkleRoot);
+    }
+
+    // When `window.locked` is true, claims are blocked for that window.
+    function setWindowLock(uint256 windowIndex, bool lockValue) external nonReentrant() onlyOwner {
+        merkleWindows[windowIndex].locked = lockValue;
+        emit SetWindowLock(msg.sender, windowIndex, lockValue);
+    }
+
+    // Overwrite merkle root for specified window index. Owner must deposit `totalRewardsDistributed`
+    // into contract.
+    // TODO: Should we require that `windowIndex <= lastSeededIndex` or should this function just be
+    // a generalized `setWindowMerkleRoot` whereby the owner can set a merkle root for any index?
+    function resetWindowMerkleRoot(
+        uint256 windowIndex,
+        uint256 totalRewardsDistributed,
+        uint256 windowStart,
+        address rewardToken,
+        bytes32 merkleRoot
+    ) external nonReentrant() onlyOwner {
+        _seedWindow(windowIndex, totalRewardsDistributed, windowStart, rewardToken, merkleRoot);
+    }
```

Comment:
Do we need this function?

In my mind, it seems bad to be able to overwrite a merkel root after claims have already been submitted. The mapping can't be deleted or overwritten, so in the new merkel root, those who have already claimed would still be marked "claimed". There would be no way to distribute to them in the updated root.

I could see two potential replacements for this functionality:
1. Lock + withdraw.
2. Delete + withdraw.

Curious to hear your thoughts?

Reply:
>  those who have already claimed would still be marked "claimed". There would be no way to distribute to them in the updated root.

Great point, although we could just store the `claimed` mapping inside the Window struct to reset it on this call.

I do think that overwriting merkle roots introduces a lot of uncertainty, in practice the owner might just find it safer to lock a Window, withdraw remaining rewards, and set a new Window.

I'm OK to remove this functionality as it doesn't seem vital. @chrismaree wdyt?

Reply:
thanks matt for bringing this up. it is wierd that you can just overwrite the root, especially seeing the examples you provide. would feel more comfortable with lock, withdraw and create.  maybe we add a seperate cleanup operation that will remove locked or expired windows/claims.

Reply:
I think Delete + withdraw might be the best because we don't have to store the `locked` variable on chain.

Reply:
@chrismaree wdyt

Reply:
@nicholaspai 

> although we could just store the claimed mapping inside the Window struct to reset it on this call.

Sadly, in solidity, there is no way to delete an entire mapping wholesale in solidity. See the docs here: https://docs.soliditylang.org/en/v0.6.12/security-considerations.html#clearing-mappings.

Reply:
Can you totally delete a struct that contains no mappings? That's what we're doing now in the latest code

Reply:
@nicholaspai yes, as far as I'm aware, normal structs containing only primitives (or nested primitives) can be deleted. I know mappings cannot be deleted. I forget whether `delete` works on arrays.